Fork TDigest library

kkrik-es opened the pull request:
The [TDigest](https://github.com/tdunning/t-digest) library is used for percentile aggregations in ES. The library contains two underlying implementations: (a) one using an AVL tree to maintain a sorted set of centroids and (b) one using pre-allocated arrays of centroids that get sorted and merged repeatedly as more samples are added. ES uses (a) by default, but 
[early results](https://github.com/elastic/elasticsearch/pull/35182#issuecomment-1514562336) show that switching to the merging implementation can offers speedups exceeding 10x in percentile calculations.

The currently used library (v.3.2) has bugs in the merging implementation that were later addressed, and the latest available version (v.3.3) provides inaccurate percentile calculations even for simple median calls. We are therefore forking the latest version and addressing these shortcomings for both implementations, before switching to the merging implementation. This PR includes fixes for MergingDigest, but keeps using AVLTreeDigest in TDigestState. The switch to MergingDigest will happen in a subsequent PR, to better track the impact of each change in isolation.

 Major changes compared to the initial implementation:
  - Quantile calculation uses interpolation between centroids instead of rounding, for singleton centroids
  - Quantile calculation properly tracks the centroid range (off-by-one error fixed)
  - Smaller sizes for weight and mean buffer arrays in MergingDigest, to reduce memory pressure
  - Centroid retrieval in MergingDigest triggers merging (if needed), as it is used when writing to StreamOutput
  - Call compress() on TDigestState objects when passed to aggregation objects, so that the former are immutable throughout the lifetime of the latter - needed for thread safety
  - No tracking of added data, other than through centroids
  - Removal of misfiring asserts and dead code, esp. around Serialization
  - Style changes to match ES codebase requirements

The top 2 changes lead to slightly different results, compared to the previous implementation from v.3.2. Still, they lead to more accurate quantile calculations. The change is thus deemed acceptable and will be documented in the release notes.

kkrik-es committed:
Initial import for TDigest forking.

kkrik-es committed:
Fix MedianTest.

More work needed for TDigestPercentile*Tests and the TDigestTest (and
the rest of the tests) in the tdigest lib to pass.

kkrik-es committed:
Fix Dist.

kkrik-es committed:
Fix AVLTreeDigest.quantile to match Dist for uniform centroids.

kkrik-es committed:
Fix `MergingDigest.quantile` to match `Dist` on uniform distribution.

elasticsearchmachine commented:
Hi @kkrik-es, I've created a changelog YAML for you.

kkrik-es committed:
Update docs/changelog/96086.yaml

kkrik-es committed:
Add merging to TDigestState.hashCode and .equals.

Remove wrong asserts from tests and MergingDigest.

kkrik-es committed:
Fix style violations for tdigest library.

kkrik-es committed:
Fix typo.

kkrik-es committed:
Fix more style violations.

kkrik-es committed:
Fix more style violations.

kkrik-es committed:
Fix remaining style violations in tdigest library.

kkrik-es committed:
Update results in docs based on the forked tdigest.

kkrik-es committed:
Fix YAML tests in aggs module.

kkrik-es committed:
Fix YAML tests in x-pack/plugin.

kkrik-es committed:
Skip failing V7 compat tests in modules/aggregations.

kkrik-es committed:
Fix TDigest library unittests.

Remove redundant serializing interfaces from the library.

kkrik-es committed:
Remove YAML test versions for older releases.

These tests don't address compatibility issues in mixed cluster tests as
the latter contain a mix of older and newer nodes, so the output depends
on which node is picked as a data node since the forked TDigest library
is not backwards compatible (produces slightly different results).

kkrik-es committed:
Fix test failures in docs and mixed cluster.

kkrik-es committed:
Reduce buffer sizes in MergingDigest to avoid oom.

kkrik-es committed:
Exclude more failing V7 compatibility tests.

martijnvg commented on x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/notification/email/HtmlSanitizerTests.java:
```diff
@@ -82,7 +82,7 @@ public void testAllowStyles() {
     public void testDefaultFormattingAllowed() {
         String html = "<b></b><i></i><s></s><u></u><o></o><sup></sup><sub></sub><ins></ins><del></del><strong></strong>"
-            + "<strike></strike><tt></tt><code></code><big></big><small></small><span></span><br /><em></em><hr />";
+            + "<strike></strike><code></code><code></code><big></big><small></small><span></span><br /><em></em><hr />";
```
Why is this change required?

martijnvg commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/TDigestState.java:
```diff
@@ -7,19 +7,18 @@
+ * Extension of {@link org.elasticsearch.tdigest.TDigest} with custom serialization.
  */
-public class TDigestState extends AVLTreeDigest {
+public class TDigestState extends MergingDigest {
```
So the wire format between avl and merging can stay the same?

kkrik-es replied:
Right, serialization is handled by this class so there are no compatibility issues - other than the changes in quantile calculations.

martijnvg commented on qa/mixed-cluster/build.gradle:
```diff
@@ -63,5 +63,19 @@ BuildParams.bwcVersions.withWireCompatible { bwcVersion, baseName ->
       dependsOn "${baseName}#mixedClusterTest"
     }
+
+    tasks.named("${baseName}#mixedClusterTest").configure {
```
I think we should use skip versions instead? I think there are other qa tests that reuse the rest api spec tests too.
This should work, since we test new behaviour and not old behavior.

martijnvg replied:
We should also ensure that at least a base percentile and percentile_ranks tests are ran in mixed cluster tests.

mark-vieira replied:
Looks like we're doing both. I saw a number of ml tests that are using `skip` in the YAML. We should align on one or the other.

martijnvg commented on benchmarks/src/main/java/org/elasticsearch/benchmark/tdigest/ApproxLogBench.java:
```diff
@@ -0,0 +1,113 @@
+ * computing the asin function involved in the merge is expensive. This argues for collecting more samples
+ * before sorting and merging them into the digest.
+ */
+@BenchmarkMode(Mode.AverageTime)
```
Should we fork all these micro benchmarks?

kkrik-es replied:
I kept just one for the tdigest and one for sorting, in case they come handy.

kkrik-es committed:
Update results for JdbcCsvSpecIT tests.

kkrik-es committed:
Update results for JdbcDocCsvSpecIT tests.

kkrik-es committed:
Revert unrelated change.

kkrik-es committed:
More test fixes.

kkrik-es committed:
Use version skips instead of blacklisting in mixed cluster tests.

kkrik-es commented:
@elasticsearchmachine run elasticsearch-ci/part-3

kkrik-es committed:
Switch TDigestState back to AVLTreeDigest.

kkrik-es committed:
Update docs and tests with AVLTreeDigest output.

kkrik-es committed:
Update flaky test.

kkrik-es committed:
Remove dead code, esp around tracking of incoming data.

kkrik-es committed:
Update docs/changelog/96086.yaml

elasticsearchmachine commented:
Hi @kkrik-es, I've updated the changelog YAML for you.

kkrik-es committed:
Delete docs/changelog/96086.yaml

kkrik-es commented:
@elastic/es-delivery can you please take at look at this change? This is related to [issue 368](https://github.com/elastic/open-source/issues/368).

kkrik-es committed:
Remove explicit compression calls.

This was added to prevent concurrency tests from failing, but it leads
to reduces precision. Submit this to see if the concurrency tests are
still failing.

kkrik-es committed:
Merge remote-tracking branch 'upstream/fix/95903' into fix/95903

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticsearchmachine commented:
Pinging @elastic/es-delivery (Team:Delivery)

kkrik-es committed:
Revert "Remove explicit compression calls."

This reverts commit 5352c96f656b9cdc63f0f673c983150bea35c7ed.

kkrik-es committed:
Remove explicit compression calls to MedianAbsoluteDeviation input.

mark-vieira commented on libs/tdigest/build.gradle:
```diff
@@ -0,0 +1,41 @@
+ * under the License.
+ */
+apply plugin: 'elasticsearch.build'
+apply plugin: 'elasticsearch.publish'
```
Keep in mind that once this is merged, we'll also need to open a pull request to the release manager to ensure this new artifact is published as part of our release process.

kkrik-es replied:
Thanks, is there documentation or some example about what the PR should contain?

mark-vieira replied:
Here's an example PR: https://github.com/elastic/infra/pull/38270/files

mark-vieira reviewed (commented):
I don't have much insight as to the implications of this change but it seems we've changed the behavior of this library as part of the fork, as a result we've had to do a few things:

1. We've updated a lot of our REST tests. My naive interpretation of this is that this means folks are getting a different result for the same query. Is that true? Would we then consider this a breaking change?
2. We've disabled a lot of mixed cluster tests due to complexities around dealing with (1). Are we concerned about losing this coverage?

kkrik-es commented:
@mark-vieira thanks for the super prompt response. You're absolutely correct about disabling tests. I've actually updated the mixed cluster config to remove blacklisting, you somehow got an outdated version of the gradle file. The only case where the tests are disabled is the V7 compatibility tests - this change doesn't affect the transport protocol so this should be safe (?).

With regard to whether this is a breaking change, we indeed produce slightly different results but this is inherently part of percentile calculations - it's [already documented](https://www.elastic.co/guide/en/elasticsearch/reference/8.7/search-aggregations-metrics-percentile-aggregation.html#search-aggregations-metrics-percentile-aggregation-approximation) that percentiles are approximate and non-deterministic, so there should be no expectations about result stability even within the same release. Still, I think it makes sense to include a note in the release notes about this change - along with an anticipated boost in performance, hopefully.

kkrik-es committed:
Add unittests for AVL and merging digest accuracy.

kkrik-es committed:
Fix spotless violations.

kkrik-es commented:
I discussed this with @martijnvg, they should be fine but Martijn can correct me.

martijnvg commented:
The tdigest library is used by the `percentile` and `percentile_ranks` aggregation. The tdigest library has minor changes between 3.2 and the forked version, but as far as I understand the core algorithm hasn't changed. However we [document](https://www.elastic.co/guide/en/elasticsearch/reference/8.8/search-aggregations-metrics-percentile-aggregation.html#search-aggregations-metrics-percentile-aggregation-approximation) that the result of these aggregations are approximate and non deterministic. This is why I think this isn't really a breaking change.

A lot of our tests are testing these aggregations with a tiny data set (a few documents) and in order for tdigest to produce meaningful results a more substantial data set is needed. So I think these tests aren't realistic. This is just the way we have been testing percentile/percentile_ranks aggregations. Now that we upgrade the tdigest library a lot of the assertions are slightly different, and this is why the skip version sections have been added. I _do_ think we need to test pre-8.9 mixed cluster behaviour, so maybe we need to add a few yaml tests that don't care about the actual percentile results, but ensure that percentile aggregations can operate without failing in a mixed cluster. Or alternatively, we remove the skip versions sections and assert that the percentile results are between a certain range? (using `close_to` or `lte`/`get` yaml assertion)

mark-vieira commented:
> I _do_ think we need to test pre-8.9 mixed cluster behaviour, so maybe we need to add a few yaml tests that don't care about the actual percentile results, but ensure that percentile aggregations can operate without failing in a mixed cluster. Or alternatively, we remove the skip versions sections and assert that the percentile results are between a certain range? (using `close_to` or `lte`/`get` yaml assertion)

👍

kkrik-es committed:
Delete redundant tests and benchmarks.

kkrik-es committed:
Fix spotless violation.

kkrik-es committed:
Use the old implementation of AVLTreeDigest.

The latest library version is 50% slower and less accurate, as verified
by ComparisonTests.

kkrik-es committed:
Merge branch 'main' into fix/95903

kkrik-es committed:
Update docs with latest percentile results.

kkrik-es committed:
Update docs with latest percentile results.

kkrik-es committed:
Merge branch 'main' into fix/95903

kkrik-es committed:
Remove repeated compression calls.

kkrik-es committed:
Update more percentile results.

kkrik-es committed:
Use approximate percentile values in integration tests.

This helps with mixed cluster tests, where some of the tests where
blocked.

kkrik-es commented:
Sounds good, I updated the tests to use approximate values and duplicated a single test per file to use exact matches and version skips, to cover both cases.

kkrik-es commented:
@elasticsearchmachine run elasticsearch-ci/bwc

kkrik-es committed:
Fix expected percentile value in test.

martijnvg commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/TDigestPercentilesAggregatorTests.java:
```diff
@@ -100,14 +100,14 @@ public void testSomeMatchesNumericDocValues() throws IOException {
-            assertEquals("1.0", tdigest.percentileAsString(25));
-            assertEquals(0.0d, tdigest.percentile(1), 0.0d);
-            assertEquals("0.0", tdigest.percentileAsString(1));
+            assertEquals(4.0d, tdigest.percentile(75), 0.01d);
```
Why not keep checking 88, 33, 35 and 1 percentiles?

kkrik-es replied:
I think these were arbitrarily selected to produce round values.

martijnvg commented on x-pack/plugin/build.gradle:
```diff
@@ -87,6 +87,15 @@ tasks.named("yamlRestTestV7CompatTest").configure {
     'aggregate-metrics/90_tsdb_mappings/aggregate_double_metric with wrong time series mappings',
     'analytics/histogram/histogram with wrong time series mappings',
     'analytics/histogram/histogram with time series mappings',
+    'analytics/boxplot/Basic Search',
```
Can these yaml tests be re-enabled and altered using gt/lt or close_to?

martijnvg commented on x-pack/qa/xpack-prefix-rest-compat/build.gradle:
```diff
@@ -185,3 +185,13 @@ tasks.named("yamlRestTestV7CompatTransform").configure{ task ->
+
+tasks.named("yamlRestTestV7CompatTest").configure {
+  systemProperty 'tests.rest.blacklist', [
+    'ml/evaluate_data_frame/Test outlier_detection auc_roc',
```
Can these yaml tests be re-enabled and be changed to use gt/lt or close_to?

martijnvg commented on x-pack/plugin/vector-tile/src/javaRestTest/java/org/elasticsearch/xpack/vectortile/VectorTileRestIT.java:
```diff
@@ -896,8 +896,11 @@ public void testMultiValueAgg() throws Exception {
-                    }
+                     "percents": [95, 99, 99.9],
+                     "tdigest": {
+                       "compression": 10
```
Why is compression 10 necessary here?

kkrik-es replied:
This was the test OOMing with MergingDigest. I implemented a hybrid digest that should address the memory allocation issue, reverting this.

martijnvg commented on server/build.gradle:
```diff
@@ -34,6 +34,7 @@ dependencies {
   api project(":libs:elasticsearch-plugin-api")
   api project(":libs:elasticsearch-plugin-analysis-api")
   api project(':libs:elasticsearch-grok')
+  api project(":libs:elasticsearch-tdigest")
```
Can we remove the official tdigest dependency here?

kkrik-es replied:
Done.

martijnvg commented on modules/aggregations/build.gradle:
```diff
@@ -43,7 +43,15 @@ tasks.named("yamlRestTestV7CompatTransform").configure { task ->
   task.skipTest("search.aggregation/210_top_hits_nested_metric/top_hits aggregation with sequence numbers", "#42809 the use nested path and filter sort throws an exception")
   task.skipTest("search.aggregation/370_doc_count_field/Test filters agg with doc_count", "Uses profiler for assertions which is not backwards compatible")
-
+  task.skipTest("search.aggregation/180_percentiles_tdigest_metric/Basic test", "Uses t-digest library which is not backwards compatible")
```
I think we can alter the assertions of these tests to match with 3.2 tdigets and new tdigest?

martijnvg commented on modules/aggregations/src/yamlRestTest/resources/rest-api-spec/test/aggregations/percentile_ranks_tdigest_metric.yml:
```diff
@@ -70,21 +96,21 @@ filtered:
             percentile_ranks_int:
               percentile_ranks:
                 field: int
-                values: [50]
```
Why the change from 50 to 51?

kkrik-es replied:
The value for 50 is 0, then goes to 16 for 51. I'm fine with either 50-0 or 51-16, up to you.

martijnvg reviewed (commented):
I left a few more comments around the tests.
Regarding the yaml tests, can we update the assertions of these tests to work with 3.2 tdigest and new tdigest (by using lt/gt or close_to)? Then there are just one version of each test that pass in normal mode, mixed cluster mode and rest bwc 7.x mode.

kkrik-es committed:
Revert in-place node updates in AVL tree.

Update quantile calculations between centroids and min/max values to
match v.3.2.

kkrik-es committed:
Add SortingDigest and HybridDigest.

The SortingDigest tracks all samples in an ArrayList that
gets sorted for quantile calculations. This approach
provides perfectly accurate results and is the most
efficient implementation for up to millions of samples,
at the cost of bloated memory footprint.

The HybridDigest uses a SortingDigest for small sample
populations, then switches to a MergingDigest. This
approach combines to the best performance and results for
small sample counts with very good performance and
acceptable accuracy for effectively unbounded sample
counts.

kkrik-es committed:
Remove deps to the 3.2 library.

kkrik-es commented:
As discussed offline, the v.7 compatibility tests use the version of the yaml test in the v.7 branch, so adding value ranges at head doesn't help.. I also played with updating the expected values, it's quite verbose. For instance, test `180_percentiles_tdigest_metric` requires the following entries here:

```
  // Replace percentile values that are affected by forking TDigest in v.8.9
  task.replaceValueInMatch("aggregations.percentiles_int.values.1\\.0", 2.5, "Basic test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.5\\.0", 8.5, "Basic test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.25\\.0", 38.5, "Basic test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.75\\.0", 113.5, "Basic test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.95\\.0", 143.49999999999997, "Basic test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.99\\.0", 149.5, "Basic test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.1\\.0", 2.5, "Basic test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.5\\.0", 8.5, "Basic test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.25\\.0", 38.5, "Basic test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.75\\.0", 113.5, "Basic test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.95\\.0", 143.49999999999997, "Basic test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.99\\.0", 149.5, "Basic test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.1\\.0", 2.5, "Metadata test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.5\\.0", 8.5, "Metadata test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.25\\.0", 38.5, "Metadata test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.75\\.0", 113.5, "Metadata test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.95\\.0", 143.49999999999997, "Metadata test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.99\\.0", 149.5, "Metadata test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.1\\.0", 2.5, "Metadata test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.5\\.0", 8.5, "Metadata test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.25\\.0", 38.5, "Metadata test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.75\\.0", 113.5, "Metadata test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.95\\.0", 143.49999999999997, "Metadata test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.99\\.0", 149.5, "Metadata test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.1\\.0", 2.5, "Only aggs test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.5\\.0", 8.5, "Only aggs test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.25\\.0", 38.5, "Only aggs test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.75\\.0", 113.5, "Only aggs test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.95\\.0", 143.49999999999997, "Only aggs test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.99\\.0", 149.5, "Only aggs test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.1\\.0", 2.5, "Only aggs test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.5\\.0", 8.5, "Only aggs test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.25\\.0", 38.5, "Only aggs test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.75\\.0", 113.5, "Only aggs test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.95\\.0", 143.49999999999997, "Only aggs test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.99\\.0", 149.5, "Only aggs test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.1\\.0", 52.0, "Filtered test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.5\\.0", 56.0, "Filtered test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.25\\.0", 76.0, "Filtered test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.75\\.0", 126.0, "Filtered test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.95\\.0", 146.0, "Filtered test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.99\\.0", 150.0, "Filtered test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.1\\.0", 52.0, "Filtered test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.5\\.0", 56.0, "Filtered test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.25\\.0", 76.0, "Filtered test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.75\\.0", 126.0, "Filtered test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.95\\.0", 146.0, "Filtered test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.99\\.0", 150.0, "Filtered test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.0.value", 8.5, "Non-keyed test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.1.value", 38.5, "Non-keyed test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.5\\.0", 8.5, "Explicit Percents test")
  task.replaceValueInMatch("aggregations.percentiles_int.values.25\\.0", 38.5, "Explicit Percents test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.5\\.0", 8.5, "Explicit Percents test")
  task.replaceValueInMatch("aggregations.percentiles_double.values.25\\.0", 38.5, "Explicit Percents test")
```

Since we're already testing functional bwc in the mixed cluster tests, this seems to have limited value. What do you think?

kkrik-es committed:
Remove unused licenses for tdigest.

martijnvg commented:
> Since we're already testing functional bwc in the mixed cluster tests, this seems to have limited value. What do you think?

Let's for now ignore the the 7.x rest backwards yaml tests (like you already did in this PR). I think this is just too verbose.

martijnvg commented on libs/tdigest/src/main/java/org/elasticsearch/tdigest/SortingDigest.java:
```diff
@@ -0,0 +1,143 @@
+ * It provides perfect quantile and cdf calculations and matches or exceeds the performance of MergingDigest at least for millions of
+ * samples, at the expense of allocating much more memory.
+ */
+public class SortingDigest extends TDigest {
```
I think this is a good idea 👍 . However can we add this class, `MergingTDigest` and related code via a separate change? This PR is already large and just keeping this PR to do bare minimum I think would be easier for the review process.

martijnvg commented on libs/tdigest/src/test/java/org/elasticsearch/tdigest/TDigestTests.java:
```diff
@@ -0,0 +1,859 @@
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+//import org.apache.commons.math3.distribution.AbstractRealDistribution;
```
remove commented out imports?

martijnvg commented on libs/tdigest/src/test/java/org/elasticsearch/tdigest/TDigestTests.java:
```diff
@@ -0,0 +1,859 @@
+        assertEquals(0, digest.centroids().size());
+    }
+
+    // /**
```
There are a lot of code commented out like this. Maybe just remove it or enable it?

kkrik-es replied:
Indeed, removed.

martijnvg commented on libs/tdigest/src/test/java/org/elasticsearch/tdigest/TDigestTests.java:
```diff
@@ -0,0 +1,859 @@
+    public void testSizeControl() throws ExecutionException, InterruptedException {
+        List<Callable<String>> tasks = new ArrayList<>();
+        for (final int size : new int[] { 10, 100, 1000, 10000 }) {
+            tasks.add(new Callable<>() {
```
maybe just () -> {...} ?

martijnvg commented on libs/tdigest/src/test/java/org/elasticsearch/tdigest/SortingDigestTests.java:
```diff
@@ -0,0 +1,38 @@
+
+import org.junit.BeforeClass;
+
+public class SortingDigestTests extends TDigestTests {
```
This is the slowest test suite on my machine. Looks like `testMonotonicity(...)` needed 28 seconds.

kkrik-es replied:
You need a new laptop :)

Changed the granularity to run 10x faster. It takes 0.25ms on my laptop, I can make it even faster as needed.

martijnvg replied:
I will inform the IT department :)

martijnvg reviewed (commented):
I submitted my review to soon... 
Left a few comments about testing and the sorting and hybrid implementations.
The later which I think it is a good idea, but should be implemented in a follow up PR.

martijnvg commented on libs/tdigest/src/test/java/org/elasticsearch/tdigest/DigestFactory.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * A DigestFactory is used in tests to abstract what kind of digest is being tested.
+ */
+public interface DigestFactory {
```
This interface looks unused? Maybe remove it?

kkrik-es replied:
Done.

martijnvg commented on libs/tdigest/src/test/java/org/elasticsearch/tdigest/MergingDigestTests.java:
```diff
@@ -0,0 +1,198 @@
+        }
+    }
+
+    // /**
```
remove this?

martijnvg commented on libs/tdigest/src/test/java/org/elasticsearch/tdigest/TDigestTests.java:
```diff
@@ -0,0 +1,859 @@
+ */
+public abstract class TDigestTests extends ESTestCase {
+
+    private static String digestName;
```
I think this field and the setup method can be removed because it seems unused?

martijnvg reviewed (commented):
I did another review round. Can all the `@SuppressWarnings("WeakerAccess")` be removed? Almost all methods have this annotation and I don't think it is relevant in the ES codebase?

kkrik-es committed:
Revert changes for SortingDigest and HybridDigest.

These will be submitted in a follow-up PR for enabling MergingDigest.

kkrik-es commented on libs/tdigest/src/test/java/org/elasticsearch/tdigest/TDigestTests.java:
```diff
@@ -0,0 +1,859 @@
+        assertEquals(20.0, dist.quantile(1), 0);
+    }
+
+    public void testSizeControl() throws ExecutionException, InterruptedException {
```
Indeed, I wanted to have another pass on the remaining tests to remove the ones that don't help with testing coverage.

martijnvg commented on libs/tdigest/src/main/java/org/elasticsearch/tdigest/Histogram.java:
```diff
@@ -0,0 +1,90 @@
+    // exposed for testing
+    abstract double lowerBound(int k);
+
+    abstract void add(Iterable<Histogram> others);
```
👍

kkrik-es committed:
Remove unused Histogram classes and unit tests.

Delete dead and commented out code, make the remaining tests run
reasonably fast. Remove unused annotations, esp. SuppressWarnings.

kkrik-es committed:
Remove Comparison class, not used.

kkrik-es committed:
Merge branch 'main' into fix/95903

kkrik-es commented:
@elasticsearchmachine run elasticsearch-ci/full-bwc

kkrik-es commented:
@elasticsearchmachine run elasticsearch-ci/full-bwc

jpountz commented on libs/tdigest/src/main/java/org/elasticsearch/tdigest/AVLTreeDigest.java:
```diff
@@ -0,0 +1,441 @@
+
+        // weightSoFar represents the total mass to the left of the center of the current node.
+        // If there's a singleton on the left boundary, there's no more mass on the left.
+        double weightSoFar = (currentWeight == 1) ? 0 : currentWeight / 2.0;
```
I can't manage to convince myself that this is correct, it feels like it should be `currentWeight/2` even when `currentWeight` is 1, given how we then sum it up with `dw`? (I could be wrong, I haven't looked at this in a very long time)

kkrik-es replied:
My understanding is that every centroid divides its weight between left and right, so about half of the samples it represents are to its left and half to its right. However, if there's a singleton on the leftmost centroid, there should be no other weight to its left. I think this is what weightSoFar represents, but I may be wrong here. `dw` extrapolates between the first and the second centroid, so the first centroid's weight (1) will still be taken into account? 

I also tried to match the results with `Dist.quantile` that uses a sorted array of all the samples. This is how I noticed this discrepancy and added this as a "fix".

What do you think? Happy to revert this if I misunderstood how it works.

kkrik-es replied:
Thanks Adrien. Discussed offline, this is a very interesting use case so I added a unit test to cover it.

The existing implementation demonstrates a non-monotonic quantile function here, a big red flag. Reverting to the old version (matching 3.2) leads to inaccuracies when the digest contains only singleton centroids, but this is preferable and a simpler solution. The inaccuracies for the singleton-only case will be mitigated when we switch to a hybrid digest that will use a sorted array of values for small sample populations.

I applied your suggestion, resolving the thread. Please feel free to reopen or add further comments.

jpountz commented on libs/tdigest/src/main/java/org/elasticsearch/tdigest/AVLTreeDigest.java:
```diff
@@ -0,0 +1,441 @@
+                return min + index * (values.mean(values.next(currentNode)) - min);
+            }
+            // Interpolate between min and first mean, otherwise.
+            return (min * index + values.mean(currentNode) * (weightSoFar - index)) / weightSoFar;
```
should we use the `weightedAverage` helper method here too for consistency?

kkrik-es replied:
Totally, updated here and in `MergingDigest`.

jpountz commented on libs/tdigest/src/test/java/org/elasticsearch/tdigest/BigCount.java:
```diff
@@ -0,0 +1,74 @@
+
+import org.elasticsearch.test.ESTestCase;
+
+public abstract class BigCount extends ESTestCase {
```
Should this be suffixed by `Tests` to meet naming conventions?

kkrik-es replied:
Done.

jpountz reviewed (commented):
I only looked at AVLTreeDigest, which is the bit I'm familiar with. One thing looks a bit fishy, but other than that it looks good to me.

kkrik-es committed:
Small fixes.

kkrik-es committed:
Merge branch 'main' into fix/95903

martijnvg reviewed (approved):
I think is ready to be merged - LGTM

kkrik-es committed:
Add javadoc and tests.

kkrik-es commented on benchmarks/src/main/java/org/elasticsearch/benchmark/tdigest/TDigestBench.java:
```diff
@@ -0,0 +1,132 @@
+
+    Random random;
+    TDigest tdigest;
+    // AbstractDistribution distribution;
```
Done.

kkrik-es commented on libs/tdigest/src/main/java/org/elasticsearch/tdigest/AVLGroupTree.java:
```diff
@@ -0,0 +1,269 @@
+        tree.update(node);
+    }
+
+    public void remove(int node) {
```
Indeed!

kkrik-es committed:
Remove special logic for singletons in the boundaries.

While this helps with the case where the digest contains only
singletons (perfect accuracy), it has a major issue problem
(non-monotonic quantile function) when the first singleton is followed
by a non-singleton centroid. It's preferable to revert to the old
version from 3.2; inaccuracies in a singleton-only digest should be
mitigated by using a sorted array for small sample counts.

kkrik-es committed:
Revert changes to expected values in tests.

This is due to restoring quantile functions to match head.

kkrik-es committed:
Revert changes to expected values in tests.

This is due to restoring quantile functions to match head.

kkrik-es committed:
Merge branch 'main' into fix/95903

kkrik-es committed:
Tentatively restore percentile rank expected results.

kkrik-es committed:
Use cdf version from 3.2

Update Dist.cdf to use interpolation, use the same cdf
version in AVLTreeDigest and MergingDigest.

kkrik-es committed:
Revert "Tentatively restore percentile rank expected results."

This reverts commit 7718dbba594f068913dfb7b85f488ca25e5e47bf.

kkrik-es committed:
Revert remaining changes compared to main.

kkrik-es commented:
@elasticsearchmachine run elasticsearch-ci/*

kkrik-es committed:
Revert excluded V7 compat tests.

kkrik-es committed:
Exclude V7 compat tests still failing.

mark-vieira reviewed (approved):
LGTM

kkrik-es committed:
Exclude V7 compat tests still failing.

kkrik-es committed:
Restore bySize function in TDigest and subclasses.

martijnvg commented:
nice! 🚀
